Offline installation and sharing functionality.

Prashant-Android opened the pull request:
- Implemented offline sharing capability.
- Updated relevant components to handle offline data sharing.

Prashant-Android committed:
Offline installation and sharing functionality.

Prashant-Android committed:
Offline installation and sharing functionality.

Prashant-Android committed:
Offline installation and sharing functionality.

poojaustad committed:
Installation

mikedawson commented on core/src/androidMain/kotlin/com/ustadmobile/core/domain/share/SendAppFileUseCaseAndroid.kt:
```diff
@@ -0,0 +1,40 @@
+                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
+            }
+
+            val chooserIntent = Intent.createChooser(intent, "Share App")
```
"Share App" - this needs to be added to the string xml. 

Catch: AppContext MUST be the activity, not the application context (otherwise localization will be wrong). Bind the use case in the activity, not in UstadApp.

mikedawson commented:
When the user clicks "Share the app" show a bottom sheet with choices: "Share app store link", "Send app file (APK)"

If user selects Share app store link, just use the package name to make the link

mikedawson commented:
Add a buildconfig field (e.g. buildconfig.default.properties) com.ustadmobile.googleplayurl set to the Ustad Mobile Google Play URL.

Then update app-android/build.gradle to add this to manifest placeholders

Then add this as a metadata proeprty to AndroidManifest (same as is done for other proeprties e.g. com.ustadmobile.presetlocale

Prashant-Android committed:
- Bottom sheet opens on share button click with options: 1) Share APK file, 2) Share app store link.
- Generated Play Store link for current package.
- Retrieved activity context from AbstractActivity.

mikedawson commented on app-android/src/main/res/xml/file_provider_path.xml:
```diff
@@ -15,4 +15,7 @@
         path="." />
-</paths>
\ No newline at end of file
+    <root-path
```
Why has this been added?

Prashant-Android replied:
The <root-path> element tells the FileProvider where to find the root directory of our app’s private storage. Without this, the FileProvider wouldn’t know where to look, and we wouldn’t be able to share the APK file.

mikedawson replied:
If that was the case, why did everything else work without it (e.g. opening submission assignment files)? I don't think it is needed. Please test removing it.

Prashant-Android replied:
I tested without this it's not working.

Failed to find configured root that contains / data/app/~~08Rm8hFVcAc05pWaKjKlyg==/ com.toughra.ustadmobile-08fWnSG-ZblrNg0W HQoPzQ==/base.apk

mikedawson commented on docs/source/installation.rst:
```diff
@@ -0,0 +1,86 @@
+6. Once the transfer is complete, tap on the received APK to install.
+
+.. image:: images/sharing1.jpg
+   :alt: Alternate text describing the image
```
Please set the alt text here e.g. Screenshot of (insert description of what is being shown on screenshot picture)

mikedawson replied:
This is not resolved.

mikedawson commented on core/src/commonMain/resources/MR/base/strings.xml:
```diff
@@ -4,6 +4,10 @@
     <!--suppress CheckTagEmptyBody -->
     <string name="blank" translatable="false"></string>
     <string name="login">Login</string>
+    <string name="share_app_title">Share App</string>
```
Strings names should be as per the text itself e.g. share_app not share_app_title (see coding style docs). Please make the string names match the text used for all strings added. 

Please also screenshot where the text is used and save the screenshot in translate/screenshots (file name should match the string name used in xml). Use an image editor (e.g. Gimp) to highlight the text within the screenshot (as per the other screenshots).

This is used in our translation platform so that translators can see a screenshot of the word in context so they can translate it properly.

mikedawson replied:
Should be "Share **a**pp" NOT "Share App"

mikedawson commented on core/src/commonMain/resources/MR/base/strings.xml:
```diff
@@ -4,6 +4,10 @@
     <string name="blank" translatable="false"></string>
     <string name="login">Login</string>
+    <string name="share_app_title">Share App</string>
+    <string name="share_app_description">Share App with Others</string>
```
Text should be "Share this app with others", which would make the name "share_this_app_with_others".

mikedawson commented:
@Prashant-Android please put the share option in the account screen (below 'add another account, above the version'). I can't see this option in any other app, so its hard to say exactly where to put it.

I think it makes sense to put this option in the same place where anything 'about app' is located - which is mostly in the account / profiles sections of all the other apps that I can find.

mikedawson commented on img/individual_create.svg:
These images are not part of offline sharing - they should not be in this pull request.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/settings/SettingsViewModel.kt:
```diff
@@ -117,7 +122,13 @@ class SettingsViewModel(
 
     private val settings: Settings by instance()
 
+    private val sendAppFileUseCase: SendAppFileUseCase by instance()
```
This would **crash** all non-Android platforms. This must be changed to use instanceOrNull. If it is null, then the share option should not be visible.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/domain/share/SendAppFileUseCase.kt:
```diff
@@ -0,0 +1,5 @@
+package com.ustadmobile.core.domain.share
+
+interface SendAppFileUseCase {
```
Let's rename this to ShareAppUseCase because it is now having both the send file option and share app link.

mikedawson commented on docs/source/installation.rst:
```diff
@@ -0,0 +1,86 @@
+   :width: 400px
+   :height: 700px
+
+Other Sharing Apps
```
Please add a link for offline sharing using Google Play. We don't need to copy from Google, just link eg. Title 

**Share offline using Google Play**

You can share the app offline using Google Play as per [Google Play instructions](https://support.google.com/googleplay/answer/9283534?hl=en).

Prashant-Android committed:
Added requested changes and updates to the offline sharing installation.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/accountlist/AccountListViewModel.kt:
```diff
@@ -91,7 +93,11 @@ class AccountListViewModel(
         ?.toBoolean() ?: false
 
+    private val shareAppUseCase: ShareAppUseCase? by instanceOrNull()
+    val sendAppOptionVisible = shareAppUseCase != null
```
This is **not** following MVVM pattern. 

MVVM works as per coding style / docs: the UiState object has properties that are then observed by the view. shareAppOptionVisible should be a property of the AccountListUiState - set by the viewmodel - like all other properties.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/accountlist/AccountListViewModel.kt:
```diff
@@ -143,6 +149,18 @@ class AccountListViewModel(
+        viewModelScope.launch {
+            try {
+                shareAppUseCase?.invoke(shareLink)
+            } catch (e: IllegalArgumentException) {
```
The catch clauses are exactly the same: you only need one catch e. Best practice now is to use Throwable instead of Exception.

mikedawson commented on core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/settings/SettingsViewModel.kt:
```diff
@@ -39,11 +40,14 @@ data class SettingsOfflineStorageOption(
 
 data class SettingsUiState(
-
+    val sendAppOptionVisible: Boolean = false,
```
Properties related to backup etc. are not related to this pull request, should not be here. sendAppOptionVisible is no longer used here (now it is moved to accountlist), so should also be removed from this uiState.

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/settings/SettingsScreen.kt:
```diff
@@ -97,43 +104,72 @@ fun SettingsScreen(
         UstadWaitForRestartDialog()
     }
 
+    if (shareAppOptionsVisible) {
```
Needs removed from here and should be only in accountlistscreen.

mikedawson commented on core/src/commonMain/resources/MR/base/strings.xml:
```diff
@@ -4,6 +4,10 @@
     <string name="blank" translatable="false"></string>
     <string name="login">Login</string>
+    <string name="share_app">Share app</string>
+    <string name="share_this_app_with_others">Share This App with Others</string>
```
This string (share_this_app_with_others) is no longer used, like anything else that is unused, it should be removed.

Prashant-Android committed:
Added requested changes and updates to the offline sharing installation.

Prashant-Android committed:
Added requested changes and updates to the offline sharing installation.

mikedawson committed:
Merge branch 'refs/heads/primary' into dev-offline-installation-sharing

mikedawson committed:
Revert changes to AccountListViewModel not related to pull request.

mikedawson committed:
Revert changes to SettingsScreen / ViewModel.

mikedawson committed:
Tidy AccountListScreen to follow Kotlin coding style.

mikedawson commented on docs/source/installation.rst:
```diff
@@ -0,0 +1,98 @@
+6. Follow the on-screen instructions to complete the installation.
+
+.. image:: images/installation1.jpg
+   :alt: Alternate text describing the image
```
These items still need alt text.

mikedawson committed:
Remove old string for share app (had incorrect capitalization).

Prashant-Android committed:
Added requested changes and updates to the offline sharing installation.

Prashant-Android committed:
Merge remote-tracking branch 'origin/dev-offline-installation-sharing' into dev-offline-installation-sharing

mikedawson commented on lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/accountlist/AccountListScreen.kt:
```diff
@@ -138,24 +168,34 @@ fun AccountListScreen(
             )
         }
 
+        item(key = "share_app") {
```
The share app option should only appear if uiState.showShareAppOption is true. This needs to go inside an if statement.

Right now - it will show on Desktop, but won't work there.

mikedawson reviewed (changes requested):
Show share app option needs to be within if condition to avoid it showing on unsupported platforms (e.g. desktop)

Prashant-Android committed:
fixed visibility of share app option according platform.

Prashant-Android committed:
fixed visibility of share app option according platform.

Prashant-Android committed:
updated manual documentation.

Prashant-Android committed:
updated manual documentation.
